### ISSUE TITLE
Microsoft Edge no longer works with version 1.3

### ISSUE THREAD

**Issue by stephan-schwarz** [PrivateBin/PrivateBin#468]
## Steps to reproduce
Open privatebin 1.3 on Microsoft Edge, and create a new paste.

### What happens
The screen will endlessly display "Sending paste" and the mouse pointer changes into the waiting symbol

### What should happen
A new paste should be created

## Additional information
Edge console output:
0: [object Object]
privatebin.js (880,12)

0: Object doesn't support property or method 'unshift'
privatebin.js (4242,12)

Note: Microsoft Edge worked fine with 1.2.1
Note 2: Internet Explorer (11), also doesn't seem to work anymore after upgrading from 1.2.1 to 1.3. It's slightly different, since it will only display "Loading..." and not show an editor.

## Basic information

**Browser:** Microsoft Edge

**PrivateBin version:** 1.3

I can reproduce this issue on <https://privatebin.net>: Yes / No
Yes

**Comment by elrido** [comment]
Unfortunately I don't think any of our core maintainers has Windows to debug this on. IE is definetly not supported anymore, see the appendix here: https://privatebin.info/news/v1.3-release.html

Any help would debugging this issue and suggested fixes would be appreciated.

**Comment by stephan-schwarz** [comment]
Debugging websites isn't really my specialty however if there is any way I can provide additional information feel free to ask.

**Comment by rugk** [comment]
First of all, maybe open the console (F12) and copy/screenshot it. (Generally, a screenshot could help.)

**Comment by r4sas** [comment]
>0: Object doesn't support property or method 'unshift'
privatebin.js (4242,12)

See same error, pointing on lines:
>            // encrypt message
>            await ServerInteraction.setCipherMessage(cipherMessage).catch(Alert.showEr

**Comment by stephan-schwarz** [comment]
Here's a screenshot

![2019-07-26 18_01_58-privatebin net - DevTools](https://user-images.githubusercontent.com/15363012/61965009-84a4e180-afcf-11e9-9f93-d66b24232a13.png)

**Comment by rugk** [comment]
Strange. The error talks about [unshift](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/unshift), but that is shown as supported by Edge un mdn.

**Comment by elrido** [comment]
The error message is probably a bit misleading - I guess that an object on that line isn't supported or returns null and an error is raised and the unshift is probably applied to when the showError method attempts to translate the error, but gets a null or other non-error instead. So unshift is supported, but only, for example, on arrays.

All of that said, the real error is probably somewhere in the encryption/compression. In FF or Chrome I would breakpoint those functions to single-step through the encryption process.

**Comment by rugk** [comment]
Could debug it a little and it seems it stops/freezes at the "import key" step (the breakpoint afterwards is not reached):
![grafik](https://user-images.githubusercontent.com/11966684/62109038-c1acf480-b2ab-11e9-997b-777a14cd95ed.png)

And I seem to get the error: 
[MDN says](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/importKey#Browser_compatibility) Edge does not support PBKDF2 for that.

**Comment by rugk** [comment]
Problems I see:
* Edge incompatibility, of course
* and that the Promise (thanks to await) seems to be ever pending or so and thus no error/anything at all is returned

**Comment by elrido** [comment]
Here is the bug, unresolved since 2016 - the suggested workaround using SHA1 won't even be enough to replicate SJCL level functionality which already used SHA256 for the HMAC:
https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/9259365/

**Comment by rugk** [comment]
Though I guess this hardly matters, as Edge is on the road of switching to being a Chromium-based browser, so it won't matter.

**Comment by eneerge** [comment]
Suggestion to add Edge to the "Old Browsers" category and display the same warning message.  There's probably some Edge only feature that's supported now that won't be in the future that can be used to detect it. 

On another note, Internet Explorer 11 doesn't appear to be able to run the "old browser" detection code, either, so it's not even warning the user that their browser is old.

SCRIPT1004: Expected ';'
zlib-1.2.11.js (6,11)
SCRIPT1004: Expected ';'
privatebin.js (767,15)

**Comment by rugk** [comment]
> On another note, Internet Explorer 11 doesn't appear to be able to run the "old browser" detection code, either, so it's not even warning the user that their browser is old.

Please open another issue for that. (I can reproduce it though)

> Suggestion to add Edge to the "Old Browsers" category and display the same warning message.

Good idea, but we have to make sure _not_ to blacklist Chromium Edge.

**Comment by elrido** [comment]
> > On another note, Internet Explorer 11 doesn't appear to be able to run the "old browser" detection code

Correct, because since ZeroBin days we always used an ```<!--[if IE]>``` tag including some CSS to display a special warning just for IE. Maybe that doesn't work anymore because we have CSP headers that prevent inline CSS now?

> Please open another issue for that.

Please don't bother opening a duplicate ticket - we already have discussed in #454 to fix the feature detection logic there.

> but we have to make sure not to blacklist Chromium Edge.

Agreed, as discussed in the last changes for the format change, we should prefer feature detection, not blanket browser version detection as the engines and version string formats change too frequently (looking at you, Opera).

Bottom line: If we improve the feature detection as per #454 (plan to work on that in early fall) both chrome in certain modes and Edge and IE should get the appropriate messages and this issue is resolved.

For users wishing to use those browsers, they can just use 1.2.1 servers until they can upgrade.

**Comment by rugk** [comment]
> Maybe that doesn't work anymore because we have CSP headers that prevent inline CSS now?

[IT does not support the (standard) CSP](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP#Browser_compatibility), so no.

**Comment by sebastiannielsen** [comment]
@eneerge I dont think MS Edge should be moved to old browsers until Chromium Edge is pre-shipped with the OS and also sent out via MS Update.
Before that, the preinstalled browser should be considered latest, and I think its better to fix Privatebin so it works in Edge.

**Comment by elrido** [comment]
If Edge's implementation of webcrypto can't deal with PBKDF2 as the above implies, then sorry, this can't be fixed. You will have to use the older SJCL release if you insist on / are forced to use Edge.

**Comment by sebastiannielsen** [comment]
@elrido or use something else, it doesn't need to be PBKDF2, like take a simpler algoritm like SHA256 and then just iterate it a few tousands of times. Or build a own implementation of PBKDF2, ergo dont use webcrypto but use a pure JS implementation of PBKDF2.

**Comment by rugk** [comment]
We don't build our own crazy crypto. Especially not as a workaround for _one_ browser (that is to be deprecated soon anyway) that has not implemented a proper key derivation function.

That said, there may be some instances that use a polyfill for the WebCrypto API, which may be needed for other reasons anyway, see https://github.com/PrivateBin/PrivateBin/issues/441#issuecomment-504656735, so maybe you can do the same on your own instance, if you have one.

Anyway, there is nothing in this issue we can do about, so what has previously been said is still true:
> Bottom line: If we improve the feature detection as per #454 (plan to work on that in early fall) both chrome in certain modes and Edge and IE should get the appropriate messages and this issue is resolved.

Given that seems to be the way forward, I'll close this in favor of https://github.com/PrivateBin/PrivateBin/issues/454, because that will add a "outdated browser" warning for Edge browsers, too.

**Comment by sebastiannielsen** [comment]
Also its important that issue #474, is present in the Chromium build of Edge. Propably this issue too. So its important that people can use the preshipped browser on any widely used system, thus if microsoft doesn't fix the issue in the chromium build, Privatebin needs to solve it.

**Comment by r4sas** [comment]
> Privatebin needs to solve it.

Feel free to open PR to propose code.
